### PR TITLE
Implements PortfolioBias in EWPCM and IWPCM

### LINKED ISSUE #3769: Implements Portfolio Bias in Portfolio Construction Models
#### Expected Behavior
Be able to create Portfolio Construction Models with Portfolio Bias. PCM with PortfolioBias.Long will convert short signals to flat and adjust the targets proportionally.

#### Actual Behavior
The portfolio bias is a concern of the alpha models.

#### Potential Solution
Implements logic in the EWPCM to enforce portfolio bias.

#### Reproducing the Problem
N/A

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by AlexCatarino**
#### Description
Implements `PortfolioBias` in EWPCM and IWPCM. With this new feature, these PCMs will ignore insights that do not respect the desired bias. E.g. for `PortfolioBias.Long`, on Insights with `InsightDirection.Up` will be converted into `PortfolioTarget.Quantity` greater than zero and other `InsightDirection` will result in `PortfolioTarget.Quantity` of zero.

#### Related Issue
Closes #3769 

#### Motivation and Context
Be able to create Portfolio Construction Models with Portfolio Bias.

#### Requires Documentation Change
Yes. Showcase this option.

#### How Has This Been Tested?
New regression algorithm and unit tests.

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.py [hunk 1]
```diff
@@ -101,12 +116,17 @@ def CreateTargets(self, algorithm, insights):
         for symbol, g in groupby(activeInsights, lambda x: x.Symbol):
             lastActiveInsights.append(sorted(g, key = lambda x: x.GeneratedTimeUtc)[-1])
 
-        # Determine target percent for the given insights
-        percents = self.DetermineTargetPercent(lastActiveInsights)
+        # Determine target percent for the given insights that respect the portfolio bias
+        percents = self.DetermineTargetPercent(list(filter(self.RespectPortfolioBias, lastActiveInsights)))
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.py]: nit - just checking, do we need to call `list()` here?

File: Tests/Algorithm/Framework/Portfolio/LongOnlyInsightWeightingPortfolioConstructionModelTests.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+using System.Linq;
+
+namespace QuantConnect.Tests.Algorithm.Framework.Portfolio
+{
+    [TestFixture]
+    public class LongOnlyInsightWeightingPortfolioConstructionModelTests : InsightWeightingPortfolioConstructionModelTests
+    {
+        private const double _weight = 0.01;
+
+        //public override double? Weight => Algorithm.Securities.Count == 0 ? default(double) : 1d / Algorithm.Securities.Count;
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Portfolio/LongOnlyInsightWeightingPortfolioConstructionModelTests.cs]: Love the unit tests 👍 

nit - commented line. A few of the using statements aren't required. `_weight` isn't required. Typo bellow `paramenter`. Same for `LongOnlyEqualWeightingPortfolioConstructionModelTests`. Suggest using ReSharper (or similar) to catch these nit cases before PR.

File: Tests/Algorithm/Framework/Portfolio/InsightWeightingPortfolioConstructionModelTests.cs [hunk 3]
```diff
@@ -13,311 +13,54 @@
...
-            var insights = new[] { GetInsight(Symbols.SPY, InsightDirection.Up, _algorithm.UtcTime) };
-
-            foreach (var target in _algorithm.PortfolioConstruction.CreateTargets(_algorithm, insights))
+            var direction = InsightDirection.Down;
+            if (PortfolioBias != PortfolioBias.LongShort && (int)direction != (int)PortfolioBias)
             {
-                var holding = _algorithm.Portfolio[target.Symbol];
-                holding.SetHoldings(holding.Price, target.Quantity);
-                _algorithm.Portfolio.SetCash(_startingCash - holding.HoldingsValue);
+                direction = InsightDirection.Flat;
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Portfolio/InsightWeightingPortfolioConstructionModelTests.cs]: Hm maybe could use `Direction.Up` when LongOnly bias, else this test won't really cover what its meant to? or actually even better directly use `Direction.Up` for all the cases.

File: Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.py (hunk 1 above)
Comment (AlexCatarino) [reply Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.py]: Yes. Python `[filter](https://docs.python.org/3/library/functions.html#filter)` returns an "iterator" that pythonnet doesn't convert into a C# List.

File: Tests/Algorithm/Framework/Portfolio/InsightWeightingPortfolioConstructionModelTests.cs (hunk 3 above)
Comment (AlexCatarino) [reply Tests/Algorithm/Framework/Portfolio/InsightWeightingPortfolioConstructionModelTests.cs]: It is right. When the `PortfolioBias` and `InsightDirection` don't meet (`PortfolioBias.LongOnly` and `InsightDirection.Down`), the `InsightDirection` is `InsightDirection.Flat`. We are creating insights with `InsightDirection.Down` in that test.

File: Tests/Algorithm/Framework/Portfolio/EqualWeightingPortfolioConstructionModelTests.cs [hunk 4]
```diff
@@ -44,7 +44,7 @@ public class EqualWeightingPortfolioConstructionModelTests
         public virtual PortfolioBias PortfolioBias => PortfolioBias.LongShort;
 
         [TestFixtureSetUp]
-        public void SetUp()
+        public virtual void SetUp()
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Portfolio/EqualWeightingPortfolioConstructionModelTests.cs]: Wondering why you made this guy virtual, I think no body is overriding him?

File: Algorithm.Framework/Portfolio/InsightWeightingPortfolioConstructionModel.cs [hunk 5]
```diff
@@ -54,7 +54,7 @@ public static InsightWeightingPortfolioConstructionModel LongOnly(TimeSpan timeS
         /// Creates a new instance of <see cref="InsightWeightingPortfolioConstructionModel"/> with Long portfolio bias.
         /// </summary>
         /// <param name="resolution">Rebalancing frequency</param>
-        public static InsightWeightingPortfolioConstructionModel LongOnly(Resolution resolution = Resolution.Daily)
+        public new static InsightWeightingPortfolioConstructionModel LongOnly(Resolution resolution = Resolution.Daily)
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Portfolio/InsightWeightingPortfolioConstructionModel.cs]: Nice catch 👍

File: Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.cs [hunk 6]
```diff
@@ -92,6 +124,16 @@ public virtual bool ShouldCreateTargetForInsight(Insight insight)
             return result;
         }
 
+        /// <summary>
+        /// Method that will determine if a given insight respects the portfolio bias
+        /// </summary>
+        /// <param name="insight">The insight to create a target for</param>
+        /// <returns>True if the insight respects the portfolio bias</returns>
+        public bool RespectPortfolioBias(Insight insight)
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.cs]: Why make `RespectPortfolioBias` public and not protected?

**Comment by jaredbroad** [comment]
Looking forward here we also have issue #https://github.com/QuantConnect/Lean/issues/3986.

So we have 2-3 parameters; rebalance frequency, minimum order size, and now portfolio bias.

I don't think we should use pseudo "extension methods" like this as the pattern gets messy and unpredictable. 

Perhaps we could consider a pattern of constructor args for all PCM's (like how all indicators are aligned)?

**Comment by jaredbroad** [comment]
Other concepts to consider: Percent deviation from the equal weighting for PCM 

https://www.quantconnect.com/forum/discussion/7182/semi-constant-rebalanced-portfolio-scrp/p1
>absolute sum of current portfolio weights diverges from the absolute sum of given portfolio weights (at initialization) more than the pre-defined threshold e.g. 10%

MinimumOrderSize which is execution concern,
https://github.com/QuantConnect/Lean/issues/3986

Both should reduce churn in different ways.

**Comment by jaredbroad** [comment]
And if we add to these models we should add to all of them simultaneously and keep the API constructors consistent between then all. Then likely it'll consolidate to the base class -- then it'll be just 3 lines of code in each consumer to reuse that base class code.

File: Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.cs [hunk 7]
```diff
@@ -131,12 +173,17 @@ public override IEnumerable<IPortfolioTarget> CreateTargets(QCAlgorithm algorith
 
             var errorSymbols = new HashSet<Symbol>();
 
-            // Determine target percent for the given insights
-            var percents = DetermineTargetPercent(lastActiveInsights);
+            // Determine target percent for the given insights that respect the portfolio bias
+            var percents = DetermineTargetPercent(lastActiveInsights.Where(RespectPortfolioBias).ToList());
```
Comment (Martin-Molinero) [inline Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.cs]: Could we directly apply the `.Where(RespectPortfolioBias)` when creating `lastActiveInsights` instead?

**Pull request closed without merging**